persistent/Database/Persist/Quasi.hs

```diff
@@ -541,31 +539,46 @@ parse :: PersistSettings -> Text -> [EntityDef]
 parse ps = maybe [] (parseLines ps) . preparse
 
 preparse :: Text -> Maybe (NonEmpty Line)
-preparse =
-    NEL.nonEmpty
-        . removeSpaces
-        . filter (not . empty)
-        . map tokenize
-        . T.lines
+preparse txt = do
+    lns <- NEL.nonEmpty (T.lines txt)
+    NEL.nonEmpty $ mapMaybe parseLine (NEL.toList lns)
+
+-- TODO: refactor to return (Line' NonEmpty)
+parseLine :: Text -> Maybe Line
+parseLine txt =
+    case tokenize txt of
+      [] ->
+          Nothing
+      toks ->
+          pure $ Line (parseIndentationAmount txt) toks
 
 -- | A token used by the parser.
-data Token = Spaces !Int   -- ^ @Spaces n@ are @n@ consecutive spaces.
-           | Token Text    -- ^ @Token tok@ is token @tok@ already unquoted.
+data Token = Token Text    -- ^ @Token tok@ is token @tok@ already unquoted.
            | DocComment Text -- ^ @DocComment@ is a documentation comment, unmodified.
   deriving (Show, Eq)
 
+tokenText :: Token -> Text
+tokenText tok =
+    case tok of
+        Token t -> t
+        DocComment t -> "-- | " <> t
+
+parseIndentationAmount :: Text -> Int
+parseIndentationAmount txt =
+    let (spaces, _) = T.span isSpace txt
+     in T.length spaces
+
 -- | Tokenize a string.
 tokenize :: Text -> [Token]
 tokenize t
     | T.null t = []
-    | "-- | " `T.isPrefixOf` t = [DocComment t]
+    | "-- | " `T.isPrefixOf` t = [DocComment (T.drop 5 t)]
```

Comment:
Oh. this is a breaking change, actually - the original code has the documentation comments *unmodified*, eg still carrying the `-- | ` stuff.

Reply:
Yeah, this isn't so great but this function here returns the `-- |` to the `Text` value representation of `DocComment`:

https://github.com/yesodweb/persistent/blob/10bc6c5868222e2ca4b3b97f672b5339aadcf3bd/persistent/Database/Persist/Quasi.hs#L560-L564

IE when it gets sent back into the `EntityDef` here

https://github.com/yesodweb/persistent/blob/10bc6c5868222e2ca4b3b97f672b5339aadcf3bd/persistent/Database/Persist/Quasi.hs#L961

`lineText` ultimately calls `tokenText` to convert the line into text for those `EntityDef` fields - there is a bit more detail on the PR description about this. I don't think it is amazing, but the state of this currently isn't too great either, and I think this change is a step in the right direction and certainly is making use of the types more rather than passing around `Text` values.

I'm happy to update/write any tests to assert nothing has broken, as this should not break anything, my intention is to maintain the existing functionality here. I have obviously had to remove/update some tests due to changing types, and removing some functions, and I've tried to update them but unfortunately it isn't the cleanest diff so that may not be so obvious (and there may be some flaws in how I've changed the tests possibly?).

Looking through the tests again, I guess we have nothing that asserts what should appear on the `EntityDef` at the end of this process (AFAICT), perhaps I could add some along with this PR to assert that that behaviour has been maintained?

Reply:
That's a good idea :)

Reply:
Ok so I've added some tests for `parse`, I've tried to cover off most variations of comments in there, but if there is a particular entity definition you want me to check (or some use case I've missed out or something), then I can always add that to these tests. You can see now though that it is retaining the `-- | ` in the values passed back to the entity def in these tests:

https://github.com/yesodweb/persistent/blob/ddb7b0afde31f29aee989bb29be2aa407f760c2c/persistent/test/main.hs#L283-L286

https://github.com/yesodweb/persistent/blob/ddb7b0afde31f29aee989bb29be2aa407f760c2c/persistent/test/main.hs#L351-L354

This field here, however, does strip out the `-- | `:

https://github.com/yesodweb/persistent/blob/ddb7b0afde31f29aee989bb29be2aa407f760c2c/persistent/test/main.hs#L361-L364

But this is the case in `master`, this function here is loading the comment into `LinesWithComments`:

https://github.com/yesodweb/persistent/blob/03e794f618b9b8daa5ae9c3a4edfe792642df55c/persistent/Database/Persist/Quasi.hs#L698-L701

Where `isComment` is stripping out the `-- | ` portion:

https://github.com/yesodweb/persistent/blob/03e794f618b9b8daa5ae9c3a4edfe792642df55c/persistent/Database/Persist/Quasi.hs#L661-L663

As a way of testing the tests, I cherry picked these onto `master` and ran them to assert that they assert the same functionality as what is in `master` currently. Hope this is sufficient to cover off that this change is not having any detrimental impact, if there is something I have missed from the above that would be good to check then I can always add more checks here :+1:

Reply:
oh man i never even thought to test that the comments weren't being picked up by attrs and extras! That feels like a silly mistake on my part.

Great investigation, thanks! I'm happy with this.